File: 23_设计类/设计版本控制(可持久化)/可持久化数组/PersistentQueue/Suspension.go
package main

import "fmt"

func main() {
	s := NewSuspensionWith(1)
	fmt.Println(s.Resolve())
	s = NewSuspensionWith(func() interface{} { return 2 })
	fmt.Println(s.Resolve())
	s = NewSuspension()
	fmt.Println(s.Resolve())
}

// 惰性求值.
type Suspension struct {
	x        interface{}
	resolved interface{}
}

func NewSuspension() *Suspension {
	return &Suspension{}
}

func NewSuspensionWith(x interface{}) *Suspension {
	return &Suspension{x: x}
}

func (s *Suspension) Resolve() interface{} {
	if s.resolved == nil {
		if f, ok := s.x.(func() interface{}); ok {
			s.resolved = f()
		} else {
			s.resolved = s.x
		}
	}
	return s.resolved
}

File: 23_设计类/设计版本控制-可持久化/可持久化数组/PersistentQueue/Suspension.ts
/**
 * 惰性求值.
 */
class Suspension<S> {
  private readonly _x: S | (() => S) | undefined = undefined
  private _resolved: S | undefined = undefined

  constructor(x?: S | (() => S)) {
    this._x = x
  }

  resolve(): S | undefined {
    // eslint-disable-next-line eqeqeq
    if (this._resolved == undefined) {
      this._resolved = typeof this._x === 'function' ? (this._x as () => S)() : this._x
    }
    return this._resolved
  }
}

export { Suspension }

if (require.main === module) {
  const s = new Suspension(() => Date.now())
  console.log(s.resolve())
}
